packages/components/src/text-input/hooks/use-text-input-state.js

```diff
@@ -0,0 +1,494 @@
+/**
+ * External dependencies
+ */
+import { isNil, noop } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import {
+	useCallback,
+	useEffect,
+	useMemo,
+	useRef,
+	useState,
+} from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { useBaseDragHandlers } from './use-text-input-state-utils';
+import { isValueNumeric } from '../../utils/values';
+import { add, roundClampString, subtract } from '../../utils/math';
+import { mergeEventHandlers } from '../../utils/events';
+import { normalizeArrowKey } from '../../utils/keyboard';
+import { useControlledValue, usePropRef } from '../../utils';
+
+/**
+ * @template T
+ * @param {Object} props
+ * @param {T} [props.value]
+ * @return {Array<any|Function>} The controlled value and the value setter.
+ */
+function useCommitValue( { value } ) {
+	const [ commitValue, setCommitValue ] = useState( null );
+	const resetCommitValue = useCallback( () => setCommitValue( null ), [] );
+
+	useEffect( resetCommitValue, [ value ] );
+
+	return [ commitValue, setCommitValue, resetCommitValue ];
+}
+
+function useShiftStep( { isShiftStepEnabled = true, shiftStep = 10 } ) {
+	const [ on, setOn ] = useState( false );
+	useEffect( () => {
+		const handleOnKeyDown = ( /** @type {KeyboardEvent}} */ event ) => {
+			if ( ! isShiftStepEnabled ) {
+				return;
+			}
+			if ( event.shiftKey ) {
+				setOn( true );
+			}
+		};
+		const handleOnKeyUp = ( /** @type {KeyboardEvent}} */ event ) => {
+			if ( ! isShiftStepEnabled ) {
+				return;
+			}
+			if ( event.shiftKey ) {
+				setOn( false );
+			}
+		};
+
+		window.addEventListener( 'keydown', handleOnKeyDown );
+		window.addEventListener( 'keyup', handleOnKeyUp );
+
+		return () => {
+			window.removeEventListener( 'keydown', handleOnKeyDown );
+			window.removeEventListener( 'keyup', handleOnKeyUp );
+		};
+	}, [ isShiftStepEnabled ] );
+
+	return isShiftStepEnabled && on ? shiftStep : 1;
+}
+
+function useFocusedState( { isFocused: isFocusedProp = false } ) {
+	const [ isFocused, setFocused ] = useState( isFocusedProp );
+
+	useEffect( () => {
+		setFocused( isFocusedProp );
+	}, [ isFocusedProp ] );
+
+	return [ isFocused, setFocused ];
+}
+
+/**
+ * @param {Object} useChangeHandlersProps
+ * @param {Function} useChangeHandlersProps.onChange
+ * @return {Object} A collection of change handlers.
```

Comment:
We can remove the return type as TS will infer it, but the `Function` type is too broad, it should be something like `(value: string) => void` I think.